### DISCUSSION TITLE
creating files from file picker

### DISCUSSION THREAD

**Discussion by gdkrmr** [helix-editor/helix#9073] [Ideas]
If I type a valid path into the file that does not exist it would be nice to create that file.

**Comment by TornaxO7** [comment]
Isn't it already possible with `:open <path>` and then `:w`?

**Reply by sdoerner** [reply]
Is there a way easily reference the current buffer dir as part of this `:open` workflow?

If this was possible from file pickers, then this woudl be easy by using "file_picker_in_current_buffer_directory".

**Comment by gdkrmr** [comment]
yes, this is possible. This isn't a reason to not being able to do it from the file picker though.

**Comment by aukeroorda** [comment]
One reason could be that you accidentally create files that you don't want. For me it would feel unintuitive/unexpected to create files from a picker

**Comment by gdkrmr** [comment]
I am coming from emacs helm where this functionality is normal. It creates a buffer and the actual file is created when the buffer is saved.

**Comment by Kxnr** [comment]
This type of functionality is something I use a lot for writing notes, where I want to find a note by subject and create one if it doesn't exist. Having fuzzy matching is really helpful for this (would past me have called that note "idea for xyz", "xyz idea", or "xyz thoughts"?). I currently use a plugin that, if the number of search results for a keyword is less than some number of results, it adds "Create <search query>" as an option to the completion menu.

If this were added as a default option I would absolutely use it, but I can also see this being a good use case for the upcoming plugin system.

----

Having looked through the picker code, it seems like adding a "Create xyz" option (as an actual item in the list) to the picker would be non-trivial [^1]. The easiest way to accomplish this behavior seems like it would be adding a keybind (with associated prompt) that I'd expect to be a relatively simple plugin. I'm going to give this a go in my own branch, and would help out with a plugin for this once such things are possible.

[^1]: ~~For one, there's no obvious way to have an item that changes in the picker list, the matcher expects a static list of items and the selection code picks an index in that list.~~ (edit: that's what the `DynamicPicker` is for 😖) The "default" item would need to be a special case everywhere it might come up through shifting indices or similar. From there, there's the issue of creating a picker item from the query string (since items are not strings themselves, they implement a formatting trait that allows them to be rendered as strings), though that would also be a challenge for a plugin implementing this

**Reply by Kxnr** [reply]
prototype on a fork [here](https://github.com/Kxnr/helix/tree/default-option-in-file-picker); adding a default option ended up being more feasible than I expected
